### PR TITLE
Fix hash function generation for frozen models with unusual MRO

### PR DESCRIPTION
**Pull request by dmontagu**
Closes https://github.com/pydantic/pydantic/issues/7236

Also fixes a bug with the way attributes were read from nested typeddicts (mro was being iterated over in reverse order), and adds a related test.

Selected Reviewer: @hramezani

### CODE REVIEW

File: tests/test_types_typeddict.py [hunk 1]
```diff
@@ -905,3 +906,16 @@ class MySubTypedDict(MyMiddleTypedDict):
...
+    class A(TypedDict):
+        x = 1
+
+    class B(A):
+        x = 2
+
+    class C(B):
+        pass
+
+    assert get_attribute_from_bases(C, 'x') == 2
```
Comment (dmontagu) [inline tests/test_types_typeddict.py:921]: Currently on the `main` branch, this returns `1`

**Comment by dmontagu** [comment]
please review

File: tests/test_types_typeddict.py (hunk 1 above)
Comment (adriangb) [reply tests/test_types_typeddict.py:921]: Oh wow I’m very glad you caught this 🙈

**Review by adriangb — APPROVED** [review]
Amazing work @dmontagu !

**Pull request merged**